rust/lance-io/src/scheduler.rs

```diff
@@ -0,0 +1,577 @@
+// Copyright 2024 Lance Developers.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use arrow_array::UInt32Array;
+use arrow_buffer::{Buffer, ScalarBuffer};
+use bytes::Bytes;
+use futures::channel::oneshot;
+use futures::{FutureExt, StreamExt};
+use object_store::path::Path;
+use snafu::{location, Location};
+use std::future::Future;
+use std::mem::size_of;
+use std::ops::Range;
+use std::sync::{Arc, Mutex};
+use tokio::sync::mpsc;
+use tokio::task::JoinHandle;
+use tokio_stream::wrappers::UnboundedReceiverStream;
+use tokio_stream::Stream;
+
+use lance_core::{Error, Result};
+
+use crate::object_store::ObjectStore;
+use crate::traits::Reader;
+
+// A request for a single range of data
+struct DirectIoRequest {
+    range: Range<u64>,
+}
+
+// A request to fetch a single range of data indirectly
+struct IndirectIoRequest {
+    offsets_range: Range<u64>,
+    data_offset: u64,
+}
+
+/// A collection of requested I/O operations
+pub struct BatchRequest {
+    direct_requests: Vec<DirectIoRequest>,
+    indirect_requests: Vec<IndirectIoRequest>,
+}
+
+impl BatchRequest {
+    /// Create an empty request
+    pub fn new() -> Self {
+        Self {
+            direct_requests: Vec::new(),
+            indirect_requests: Vec::new(),
+        }
+    }
+
+    /// Add a request to read a specified range of bytes from the disk
+    pub fn direct_read(&mut self, range: Range<u64>) {
+        self.direct_requests.push(DirectIoRequest { range })
+    }
+
+    /// Add a request to read a range of bytes from disk indirectly
+    ///
+    /// This assumes that the offsets of the the desired data are
+    /// written somewhere.  It first reads those offsets and then reads
+    /// the range specified by those offsets.
+    ///
+    /// For example, if the offsets are written to the start of the file
+    ///
+    /// [0]: 100
+    /// [4]: 120
+    /// [8]: 300
+    /// [12]: 330
+    ///
+    /// And the request has offsets_range: 4..16 then it will
+    /// first read those offsets (120, 300, and 330) and then read that range
+    /// (120..330) from the disk.
+    ///
+    /// All offsets are assumed to be little endian u32 values.
```

Comment:
I feel very confused. Are we expecting to have a lot of u32 offset arrays in the new file format?

For Arrow list arrays, they use signed integers for compatibility with Java. That seems like a wise choice.

Reply:
At some point this will need to become a generic method for all the offset types.  I can probably go ahead and do that now.  I was holding off why I figured out the general shape.

I expect, pretty quickly, that we will want to use offsets that are as small as possible.  E.g. we should never need to use 64-bit offsets unless the page is bigger than 4GiB (which should only happen if an individual element was bigger than 4GiB).  This is true even if the resulting array is large_string / large_binary / whatever.  The cost of converting a 32-bit array to a 64-bit array should usually be faster than the extra I/O cost (though maybe not in all situations).

Even if we use the same format that we eventually want in Arrow that doesn't guarantee zero-copy.  Any array that spans more than one page will force a copy (I expect these will mostly be on the edges of pages).

> For Arrow list arrays, they use signed integers for compatibility with Java. That seems like a wise choice.

Java today has pretty good support for [treating signed integers as if they were unsigned integers](https://www.baeldung.com/java-unsigned-arithmetic) so I'm not sure this reasoning still holds up.  Although I don't expect we lose all that much by using signed integers.